pkg/cmd/prune-backups/prune.go

```diff
@@ -0,0 +1,226 @@
+		return nil
+	} else if r.retentionType == RetentionTypeSize {
+		return retainBySizeGb(r.maxSizeOfBackupsGb)
+	} else if r.retentionType == RetentionTypeNumber {
+		return retainByNumber(r.maxNumberOfBackups)
+	}
+
+	return nil
+}
+
+func retainBySizeGb(sizeInGb int) error {
+	folders, err := listAllBackupFolders()
+	if err != nil {
+		return err
+	}
+
+	sort.Sort(folders)
+	// we keep the latest - up to sizeInGb folders around, the remainder is deleted
+	cutOffBytes := int64(1024 * 1024 * sizeInGb)
+	accBytes := int64(0)
+	var toRemove []string
+	for _, f := range folders {
+		accBytes += f.sizeBytes
+		if accBytes > cutOffBytes {
+			toRemove = append(toRemove, path.Join(BasePath, f.name))
+		} else {
+			klog.Infof("retaining [%s], found [%d] bytes so far", f.name, accBytes)
+		}
+	}
+
+	for _, f := range toRemove {
+		klog.Infof("deleting [%s]...", f)
+		err = os.RemoveAll(f)
+		if err != nil {
+			return fmt.Errorf("error while removing [%s]: %w", f, err)
+		}
+	}
+
+	klog.Infof("pruning successful")
+	return nil
+}
+
+func retainByNumber(maxNumBackups int) error {
+	folders, err := listAllBackupFolders()
+	if err != nil {
+		return err
+	}
+
+	if len(folders) <= maxNumBackups {
+		klog.Infof("numFolders=[%d] which is smaller or equal to requested [%d], no pruning necessary", len(folders), maxNumBackups)
+		return nil
+	}
+
+	sort.Sort(folders)
+	for _, f := range folders[maxNumBackups:] {
+		bPath := path.Join(BasePath, f.name)
+		klog.Infof("deleting [%s]...", bPath)
+		err = os.RemoveAll(bPath)
+		if err != nil {
+			return fmt.Errorf("error while removing [%s]: %w", bPath, err)
+		}
+	}
+	klog.Infof("pruning successful")
+	return nil
+}
+
+func listAllBackupFolders() (backupDirStats, error) {
+	var stats []backupDirStat
+
+	dir, err := os.ReadDir(BasePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not list dir [%s]: %w", dir, err)
+	}
+
+	for _, d := range dir {
+		if !d.IsDir() {
+			continue
+		}
+
+		var dirSize int64
+		var latestModTime time.Time
+		err := fs.WalkDir(os.DirFS(path.Join(BasePath, d.Name())), ".", func(path string, d fs.DirEntry, err error) error {
+			if !d.IsDir() {
+				info, err := d.Info()
+				if err != nil {
+					return err
+				}
+				dirSize += info.Size()
+				if latestModTime.Before(info.ModTime()) {
+					latestModTime = info.ModTime()
+				}
+			}
+			return nil
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("could not recurse into dir [%s]: %w", d.Name(), err)
+		}
+
+		stats = append(stats, backupDirStat{
+			name:      d.Name(),
+			sizeBytes: dirSize,
+			modTime:   latestModTime,
+		})
+
+	}
+
+	klog.Infof("found backup folders: %v", stats)
+	return stats, nil
+}
+
+func (s backupDirStat) String() string {
+	return fmt.Sprintf("Name=[%s] SizeBytes=[%d] ModTime=[%s]", s.name, s.sizeBytes, s.modTime.String())
+}
+
+func (b backupDirStats) Len() int {
+	return len(b)
+}
+
+func (b backupDirStats) Less(i, j int) bool {
+	return b[j].modTime.Before(b[i].modTime)
```

Comment:
leaving this for discussion: WDYT about also sorting by the name? in case some filesystem doesn't have proper mod times

Reply:
Not sure if there are filesystems that won't set mtime. But our backup files names are timestamped right?
https://github.com/openshift/cluster-etcd-operator/blob/381ffb81706699cdadd0735a52f9d20379505ef7/pkg/operator/backupcontroller/backupcontroller.go#L313

So that should give us the ordering on the creation time as well. So we could do that as well if we don't find the modtime set.